pkg/controller/admissionchecks/multikueue/remoteclient.go

```diff
@@ -0,0 +1,180 @@
+/*
+Copyright 2023 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package multikueue
+
+import (
+	"context"
+
+	"k8s.io/apimachinery/pkg/api/equality"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/clientcmd"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+
+	kueue "sigs.k8s.io/kueue/apis/kueue/v1beta1"
+)
+
+type remoteController struct {
+	localClient   client.Client
+	watchCtx      context.Context
+	watchCancel   context.CancelFunc
+	remoteClients map[string]*remoteClient
+	wlUpdateCh    chan<- event.GenericEvent
+
+	// for testing only
+	restConfigFromKubeConfigOverride func([]byte) (*rest.Config, error)
+	newWithWatchOverride             func(config *rest.Config, options client.Options) (client.WithWatch, error)
+}
+
+func newRemoteController(watchCtx context.Context, localClient client.Client, wlUpdateCh chan<- event.GenericEvent) *remoteController {
+	watchCtx, watchCancel := context.WithCancel(watchCtx)
+	ret := &remoteController{
+		localClient: localClient,
+		watchCtx:    watchCtx,
+		watchCancel: watchCancel,
+		wlUpdateCh:  wlUpdateCh,
+	}
+	return ret
+}
+
+func (rc *remoteController) UpdateConfig(ctx context.Context, kubeConfigs map[string][]byte) error {
+	if rc.remoteClients == nil {
+		rc.remoteClients = make(map[string]*remoteClient, len(kubeConfigs))
+	}
+
+	for clusterName, c := range rc.remoteClients {
+		if kubeconfig, found := kubeConfigs[clusterName]; found {
+			if err := c.setConfig(kubeconfig); err != nil {
+				delete(rc.remoteClients, clusterName)
+				return err
+			}
+		} else {
+			c.watchCancel()
+			delete(rc.remoteClients, clusterName)
+		}
+	}
+	// create the missing ones
+	for clusterName, kubeconfig := range kubeConfigs {
+		if _, found := rc.remoteClients[clusterName]; !found {
+			c := newRemoteClient(rc.watchCtx, rc.localClient, rc.wlUpdateCh)
+
+			c.restConfigFromKubeConfigOverride = rc.restConfigFromKubeConfigOverride
+			c.newWithWatchOverride = rc.newWithWatchOverride
+
+			if err := c.setConfig(kubeconfig); err != nil {
+				return err
+			}
+			rc.remoteClients[clusterName] = c
+		}
+	}
+
+	return nil
+}
+
+func (cc *remoteController) IsActive() bool {
+	//TODO improve the logic
```

Comment:
There are a couple of TODO comments. Please prepare a list of follow ups and add to the PR description, so that we don't forget creating PRs/Issues out of them.